makefiles/features_check.inc.mk

```diff
@@ -1,41 +1,66 @@
 # Check if all required FEATURES are provided
 
-FEATURES_OPTIONAL_ONLY := $(sort $(filter-out $(FEATURES_REQUIRED),$(FEATURES_OPTIONAL)))
-FEATURES_OPTIONAL_USED := $(sort $(filter $(FEATURES_PROVIDED),$(FEATURES_OPTIONAL_ONLY)))
+# Receives a list of list of '|' separated features and resturns for each
+# item in list, the first match that is present in the whitelist.
+#
+# _features_one_out_of <features-list> <whitelist>
+#
+# Parameters
+#   features-list: list of list of '|' separated features
+#     e.g.: "periph_spi|periph_i2c puf_sram|periph_hwrng"
+#   whitelist: possible features to select one out of
+#     e.g.: "$(FEATURES_PROVIDED)"
+#
+# One out of Algorithm:
+#  - For each list in features-list as "item":
+#     - Store the intersection of features-whitelist and the features in
+#       "item" in "tmp"
+#     - Append "item" to "tmp" (with pipes between features, e.g.
+#       "periph_spi|periph_i2c")
+#     - Appends the first's element of "tmp" to the caller
+#   ==> If one (or more) features in whitelist is listed in item, the first
+#       one will be taken.
+#   ==> At the end of the list item itself (with pipes between features) is the
+#       last item in "tmp". If no feature in the feature-list is in whitelist
+#       this will be the only item in "tmp" and be picked
+_features_one_out_of = $(foreach item, $1, $(firstword \
+                          $(filter $(subst |, ,$(item)), $2) $(item)))
+
+# Features that are provided and not blacklisted
+FEATURES_USABLE := $(filter-out $(FEATURES_BLACKLIST),$(FEATURES_PROVIDED))
+
+# Additionally optional features due to the "one out of" dependencies,
+#   For each features list in $(FEATURES_OPTIONAL_ALT):
+#     ==> If one (or more) features is usable select the first one in list
+#     ==> If no feature is usable return the list
+FEATURES_OPTIONAL_ONE_OUT_OF := $(call _features_one_out_of,\
+                                  $(FEATURES_OPTIONAL_ALT),\
+                                    $(FEATURES_OPTIONAL) \
+                                    $(FEATURES_USABLE))
+
+FEATURES_OPTIONAL_ONLY := $(sort $(filter-out $(FEATURES_REQUIRED),\
+                            $(FEATURES_OPTIONAL) \
+                            $(FEATURES_OPTIONAL_ONE_OUT_OF)))
+FEATURES_OPTIONAL_USED := $(sort $(filter $(FEATURES_PROVIDED), \
+                            $(FEATURES_OPTIONAL_ONLY)))
```

Comment:
After a comment from Leandro offline, this should probably be:

```suggestion
# FEATURES_OPTIONAL that have not been required
FEATURES_OPTIONAL_ONLY_SO_FAR := $(sort $(filter-out $(FEATURES_REQUIRED),\
                            $(FEATURES_OPTIONAL)))
# Only add FEATURES that are usable
FEATURES_OPTIONAL_USED_SO_FAR := $(sort $(filter $(FEATURES_USABLE) \
                                    $(FEATURES_OPTIONAL_ONLY_SO_FAR)))


# Additionally optional features due to the "one out of" dependencies,
#   For each features list in $(FEATURES_OPTIONAL_ALT):
#     ==> If one (or more) features is already added to OPTIONAL_USED
          then the firs of the used ones is returned
#     ==> If one (or more) features is usable select the first one in list
#     ==> If no feature is usable return the list
FEATURES_OPTIONAL_ONE_OUT_OF := $(call _features_one_out_of,\
                                  $(FEATURES_OPTIONAL_ALT),\
                                    $(FEATURES_OPTIONAL_USED_SO_FAR) \
                                    $(FEATURES_USABLE))

# FEATURES_OPTIONAL that have not been required
FEATURES_OPTIONAL_ONLY := $(sort $(filter-out $(FEATURES_REQUIRED),\
                            $(FEATURES_OPTIONAL_ONE_OUT_OF)))
# Only add FEATURES that are usable
FEATURES_OPTIONAL_USED := $(sort $(filter $(FEATURES_USABLE) \
                                    $(FEATURES_OPTIONAL_ONLY )))
```

This way do not add to `FEATURES_OPTIONAL_USED` non-usable features, and do not add with the `FEATURES_OPTIONAL_ALT` a feature that will be added bt the OPTIONAL mechanism. In other words:

`FEATURES_OPTIONAL` has precedence over `FEATURES_OPTIONAL_ALT`